src/python.cpp

```diff
@@ -2207,3 +2207,108 @@ SEXP py_eval_impl(const std::string& code, bool convert = true) {
   return result;
 
 }
+
+// [[Rcpp::export]]
+SEXP py_convert_pandas_series(PyObjectRef series) {
+  
+  // extract dtype
+  PyObjectPtr dtype(PyObject_GetAttrString(series, "dtype"));
+  PyObjectPtr name(PyObject_GetAttrString(dtype, "name"));
+  
+  // as well as values
+  PyObjectPtr values(PyObject_GetAttrString(series, "values"));
+  
+  RObject R_obj;
+  
+  // special treatment for pd.Categorical
+  if (as_std_string(name) == "category") {
+    
+    // get actual values and convert to R
+    PyObjectPtr cat(PyObject_GetAttrString(series, "cat"));
+    PyObjectPtr codes(PyObject_GetAttrString(cat, "codes"));
+    PyObjectPtr code_values(PyObject_GetAttrString(codes, "values"));
+    RObject R_values = py_to_r(code_values, true);
+    
+    // get levels and convert to R
+    PyObjectPtr categories(PyObject_GetAttrString(dtype, "categories"));
+    PyObjectPtr category_values(PyObject_GetAttrString(categories, "values"));
+    RObject R_levels = py_to_r(category_values, true);
+    
+    // get "ordered" attribute
+    PyObjectPtr ordered(PyObject_GetAttrString(dtype, "ordered"));
+    //RObject ordered = py_to_r(ordered_, true);
+    
+    // populate integer vector to hold factor values
+    int* codes_int = INTEGER(R_values);
+    int n = Rf_length(R_values);
+    IntegerVector factor(n);
+    // values need to start at 1
+    for (int i = 0; i < n; ++i) {
+      factor[i] = codes_int[i] + 1;
+    }
+    
+    // populate character vector to hold levels
+    n = Rf_length(R_levels);
+    CharacterVector factor_levels(n);
+    for (int i = 0; i < n; ++i) {
+      factor_levels[i] = STRING_ELT(R_levels, i);
+    }
+    
+    factor.attr("class") = "factor";
+    factor.attr("levels") = factor_levels;
+    if (PyObject_IsTrue(ordered)) factor.attr("ordered") = true;
+    
+    R_obj = factor;
+    
+    // default case
+  } else {
+    
+    R_obj = py_to_r(values, series.convert());
+    
+  }
+  
+  return R_obj;
+  
+}
+
+// [[Rcpp::export]]
+SEXP py_convert_pandas_df(PyObjectRef df) {
+  
+  // pd.DataFrame.items() returns an Iterator over (column name, Series) pairs
+  PyObjectPtr items(PyObject_CallMethod(df, "items", NULL));
+  if (! PyObject_HasAttrString(items, "__next__"))
+    stop("Cannot iterate over object");
+
+  std::vector<RObject> list;
+  
+  while (true) {
+
+    // get next tuple
+    PyObjectPtr tuple(PyIter_Next(items));
+    if (tuple.is_null()) {
+      if (PyErr_Occurred())
+        stop(py_fetch_error());
+      else
+        break;
+    }
+    
+    // access Series in slot 1
+    PyObjectPtr series(PySequence_GetItem(tuple, 1));
+    Py_IncRef(series);
```

Comment:
The downside with using `Py_IncRef()` directly here is that, if a C++ exception happened to be emitted before we reached the `Py_DecRef()` call following, we'd end up leaking the object.

Do we need to call `Py_IncRef()` here? I think the object is basically protected by virtue of living within a `PyObjectPtr` already; this may be unnecessary.

Reply:
I did this in response to  a segfault I got... but now I see that if I remove those statements, I still get the segfault. It does not (or not reliably) occur after a single roundtrip, but can be produced by 

```
before <- iris
p <- r_to_py(before)

microbenchmark::microbenchmark(
  py_to_r(p),
  times = 100
)
```

I'll debug this after pushing the other changes.
Just quickly noting 

```
(gdb) bt
#0  0x00007f48963cf63b in PyObject_GetAttr () from /usr/lib64/libpython3.7m.so
...
#153 0x00007f48963df167 in PyObject_GetAttrString () from /usr/lib64/libpython3.7m.so
#154 0x00007f48b04e3366 in py_convert_pandas_series (series=...) at reticulate_types.h:38

(gdb) frame 154
#154 0x00007f48b04e3366 in py_convert_pandas_series (series=...) at reticulate_types.h:38
38	    return get();
```